File: backend/app/games/mendikot/engine.py
"""Mendikot pure game logic — Python port of the frontend engine.

Official rules: 4 players in two partnerships (seats 0/2 vs 1/3), capture the
four 10s. 3+ tens wins; 2-2 split decided by tricks; all four tens = mendikot;
all 13 tricks = whitewash. Trump is chosen secretly and revealed the first
time a player cannot follow suit.
"""
from __future__ import annotations

import random
from typing import List, Optional

from ..kachuful.deck import RANK_VALUE
from ..kachuful.types import RANKS, SUITS, Card, Suit
from .types import HandResult, MendikotState, MPlayer, PlayedCard

_SUIT_ORDER = {"S": 0, "H": 1, "C": 2, "D": 3}
_RED = {"H", "D"}


def create_deck() -> List[Card]:
    return [Card(id=f"{r}{s}", suit=s, rank=r) for s in SUITS for r in RANKS]


def shuffle(cards: List[Card]) -> List[Card]:
    out = list(cards)
    random.shuffle(out)
    return out


def sort_hand(hand: List[Card]) -> List[Card]:
    return sorted(
        hand, key=lambda c: (_SUIT_ORDER[c.suit], -RANK_VALUE[c.rank])
    )


def team_of(seat: int) -> int:
    return seat % 2


def next_seat(seat: int) -> int:
    return (seat + 1) % 4


def is_red(suit: Suit) -> bool:
    return suit in _RED


def player_at_seat(state: MendikotState, seat: int) -> MPlayer:
    return next(p for p in state.players if p.seat == seat)


def _refresh_counts(state: MendikotState) -> None:
    state.handCounts = {pid: len(cards) for pid, cards in state.hands.items()}


def create_initial_state(
    players: List[MPlayer], settings, host_id: Optional[str] = None
) -> MendikotState:
    deck = shuffle(create_deck())
    ps = sorted(players, key=lambda p: p.seat)
    hands = {}
    for i, p in enumerate(ps):
        hands[p.id] = sort_hand(deck[i * 13 : i * 13 + 13])
    state = MendikotState(
        players=ps,
        hands=hands,
        chooserSeat=0,
        leaderSeat=0,
        turnSeat=0,
        phase="choose_trump",
        message="Choosing the trump suit…",
        hostId=host_id,
    )
    _refresh_counts(state)
    return state


def legal_cards(state: MendikotState, seat: int) -> List[Card]:
    hand = state.hands[player_at_seat(state, seat).id]
    if state.leadSuit is None:
        return hand
    in_suit = [c for c in hand if c.suit == state.leadSuit]
    return in_suit if in_suit else hand


def trick_winner_seat(
    trick: List[PlayedCard], trump: Optional[Suit], lead_suit: Suit
) -> int:
    trumps = [p for p in trick if trump and p.card.suit == trump]
    pool = trumps if trumps else [p for p in trick if p.card.suit == lead_suit]
    best = pool[0]
    for p in pool[1:]:
        if RANK_VALUE[p.card.rank] > RANK_VALUE[best.card.rank]:
            best = p
    return best.seat


def active_trump(state: MendikotState) -> Optional[Suit]:
    return state.trump if state.trumpRevealed else None


def current_expected_player_id(state: MendikotState) -> Optional[str]:
    if state.phase == "choose_trump":
        return player_at_seat(state, state.chooserSeat).id
    if state.phase == "playing":
        return player_at_seat(state, state.turnSeat).id
    return None


def turn_token(state: MendikotState) -> tuple:
    return (
        state.phase,
        state.turnSeat,
        len(state.currentTrick),
        state.chooserSeat,
        state.trumpRevealed,
    )


def choose_trump(state: MendikotState, suit: Suit) -> MendikotState:
    if state.phase != "choose_trump":
        return state
    s = state.model_copy(deep=True)
    s.trump = suit
    s.phase = "playing"
    s.turnSeat = s.leaderSeat
    s.message = f"{player_at_seat(s, s.turnSeat).name} leads"
    return s


def play_card(state: MendikotState, card_id: str) -> MendikotState:
    if state.phase != "playing":
        return state
    s = state.model_copy(deep=True)
    seat = s.turnSeat
    player = player_at_seat(s, seat)
    hand = s.hands[player.id]
    card = next((c for c in hand if c.id == card_id), None)
    if card is None:
        return state
    legal = legal_cards(s, seat)
    if not any(c.id == card_id for c in legal):
        return state

    s.hands[player.id] = [c for c in hand if c.id != card_id]
    if not s.currentTrick:
        s.leadSuit = card.suit
    if (not s.trumpRevealed) and s.leadSuit and card.suit != s.leadSuit:
        s.trumpRevealed = True
        s.message = "Trump revealed!"
    s.currentTrick.append(PlayedCard(seat=seat, card=card))

    if len(s.currentTrick) < 4:
        s.turnSeat = next_seat(seat)
        _refresh_counts(s)
        return s

    winner = trick_winner_seat(s.currentTrick, active_trump(s), s.leadSuit)
    w_team = team_of(winner)
    s.tricksByTeam[w_team] += 1
    tens_here = sum(1 for t in s.currentTrick if t.card.rank == "10")
    s.tensByTeam[w_team] += tens_here
    s.lastTrickWinnerSeat = winner
    s.phase = "trick_done"
    s.message = f"{player_at_seat(s, winner).name} wins the trick"
    _refresh_counts(s)
    return s


def advance_after_trick(state: MendikotState) -> MendikotState:
    if state.phase != "trick_done":
        return state
    s = state.model_copy(deep=True)
    hands_empty = all(len(s.hands[p.id]) == 0 for p in s.players)
    s.currentTrick = []
    s.leadSuit = None
    if hands_empty:
        s.phase = "hand_over"
        s.result = compute_result(s)
        s.message = s.result.message
        return s
    s.phase = "playing"
    s.leaderSeat = s.lastTrickWinnerSeat
    s.turnSeat = s.lastTrickWinnerSeat
    s.message = f"{player_at_seat(s, s.turnSeat).name} leads"
    return s


def is_won(state: MendikotState) -> bool:
    return state.phase == "hand_over"


def compute_result(s: MendikotState) -> HandResult:
    t_a, t_b = s.tensByTeam
    tr_a, tr_b = s.tricksByTeam
    if tr_a == 13 or tr_b == 13:
        winner = 0 if tr_a == 13 else 1
        kind = "whitewash"
    elif t_a == 4 or t_b == 4:
        winner = 0 if t_a == 4 else 1
        kind = "mendikot"
    elif t_a >= 3 or t_b >= 3:
        winner = 0 if t_a >= 3 else 1
        kind = "tens"
    else:
        winner = 0 if tr_a > tr_b else 1
        kind = "tricks"

    side = "Team A" if winner == 0 else "Team B"
    messages = {
        "mendikot": f"{side} swept all four 10s — Mendikot!",
        "whitewash": f"{side} took all 13 tricks — Whitewash!",
        "tens": f"{side} captured the majority of 10s.",
        "tricks": f"10s split 2-2 — {side} won on tricks.",
    }
    return HandResult(
        winnerTeam=winner,
        kind=kind,
        tens=[t_a, t_b],
        tricks=[tr_a, tr_b],
        message=messages[kind],
    )


def suit_counts(hand: List[Card]) -> dict:
    c = {"S": 0, "H": 0, "D": 0, "C": 0}
    for card in hand:
        c[card.suit] += 1
    return c

File: src/games/mendikot/engine/engine.ts
import { createDeck, shuffle, RANK_VALUE, sortHand } from "../../kachuful/engine/deck";
import {
  HUMAN_SEAT,
  nextSeat,
  teamOf,
  type Card,
  type HandResult,
  type MendikotState,
  type MPlayer,
  type PlayedCard,
  type Suit,
} from "./types";

const BOT_NAMES = ["Rahul", "Anjali", "Vikram"];
const BOT_AVATARS = ["🦊", "🐼", "🦉"];

/** Build the 4 seated players (you bottom, partner top). */
function makePlayers(youName: string): MPlayer[] {
  const you: MPlayer = {
    id: "you",
    name: youName || "You",
    seat: HUMAN_SEAT,
    isBot: false,
    team: teamOf(HUMAN_SEAT),
    avatar: "🧑‍🎓",
  };
  const bots: MPlayer[] = [1, 2, 3].map((seat, i) => ({
    id: `bot${seat}`,
    name: seat === 2 ? `${BOT_NAMES[i]} (Partner)` : BOT_NAMES[i],
    seat,
    isBot: true,
    team: teamOf(seat),
    avatar: BOT_AVATARS[i],
  }));
  return [you, ...bots].sort((a, b) => a.seat - b.seat);
}

export function createGame(youName: string): MendikotState {
  const deck = shuffle(createDeck());
  const players = makePlayers(youName);
  const hands: Record<string, Card[]> = {};
  players.forEach((p, i) => {
    hands[p.id] = sortHand(deck.slice(i * 13, i * 13 + 13));
  });

  return {
    players,
    hands,
    trump: null,
    trumpRevealed: false,
    chooserSeat: HUMAN_SEAT,
    leaderSeat: HUMAN_SEAT,
    turnSeat: HUMAN_SEAT,
    currentTrick: [],
    leadSuit: null,
    tensByTeam: [0, 0],
    tricksByTeam: [0, 0],
    lastTrick: null,
    lastTrickWinnerSeat: null,
    phase: "choose_trump",
    message: "Choose the trump suit",
    result: null,
  };
}

function clone(s: MendikotState): MendikotState {
  const hands: Record<string, Card[]> = {};
  for (const k of Object.keys(s.hands)) hands[k] = [...s.hands[k]];
  return {
    ...s,
    players: s.players.map((p) => ({ ...p })),
    hands,
    currentTrick: s.currentTrick.map((t) => ({ ...t })),
    tensByTeam: [...s.tensByTeam] as [number, number],
    tricksByTeam: [...s.tricksByTeam] as [number, number],
    lastTrick: s.lastTrick ? s.lastTrick.map((t) => ({ ...t })) : null,
    result: s.result ? { ...s.result } : null,
  };
}

export function seatById(s: MendikotState, id: string): number {
  return s.players.find((p) => p.id === id)!.seat;
}

export function playerAtSeat(s: MendikotState, seat: number): MPlayer {
  return s.players.find((p) => p.seat === seat)!;
}

/** Cards the player at `seat` is allowed to play right now. */
export function legalCards(s: MendikotState, seat: number): Card[] {
  const hand = s.hands[playerAtSeat(s, seat).id];
  if (!s.leadSuit) return hand; // leader plays anything
  const inSuit = hand.filter((c) => c.suit === s.leadSuit);
  return inSuit.length > 0 ? inSuit : hand;
}

/** Winner seat of a (partial or full) trick, given active trump + lead suit. */
export function trickWinnerSeat(
  trick: PlayedCard[],
  trump: Suit | null,
  leadSuit: Suit
): number {
  const trumps = trick.filter((p) => trump && p.card.suit === trump);
  const pool = trumps.length > 0 ? trumps : trick.filter((p) => p.card.suit === leadSuit);
  return pool.reduce((best, cur) =>
    RANK_VALUE[cur.card.rank] > RANK_VALUE[best.card.rank] ? cur : best
  ).seat;
}

/** The active trump for winner calculations (null until revealed). */
export function activeTrump(s: MendikotState): Suit | null {
  return s.trumpRevealed ? s.trump : null;
}

export function chooseTrump(state: MendikotState, trump: Suit): MendikotState {
  if (state.phase !== "choose_trump") return state;
  const s = clone(state);
  s.trump = trump;
  s.phase = "playing";
  s.turnSeat = s.leaderSeat;
  s.message = `${playerAtSeat(s, s.turnSeat).name} leads`;
  return s;
}

/** Play one card for the player whose turn it is. Returns new state or null. */
export function playCard(state: MendikotState, cardId: string): MendikotState | null {
  if (state.phase !== "playing") return null;
  const s = clone(state);
  const seat = s.turnSeat;
  const player = playerAtSeat(s, seat);
  const hand = s.hands[player.id];
  const card = hand.find((c) => c.id === cardId);
  if (!card) return null;

  // enforce follow-suit
  const legal = legalCards(s, seat);
  if (!legal.some((c) => c.id === cardId)) return null;

  // remove from hand
  s.hands[player.id] = hand.filter((c) => c.id !== cardId);

  if (s.currentTrick.length === 0) s.leadSuit = card.suit;

  // reveal trump the first time someone can't follow suit
  if (!s.trumpRevealed && s.leadSuit && card.suit !== s.leadSuit) {
    s.trumpRevealed = true;
    s.message = `Trump revealed!`;
  }

  s.currentTrick.push({ seat, card });

  if (s.currentTrick.length < 4) {
    s.turnSeat = nextSeat(seat);
    return s;
  }

  // trick complete → resolve
  const winner = trickWinnerSeat(s.currentTrick, activeTrump(s), s.leadSuit!);
  const wTeam = teamOf(winner);
  s.tricksByTeam[wTeam] += 1;
  const tensHere = s.currentTrick.filter((t) => t.card.rank === "10").length;
  s.tensByTeam[wTeam] += tensHere;
  s.lastTrick = s.currentTrick;
  s.lastTrickWinnerSeat = winner;
  s.phase = "trick_done";
  s.message = `${playerAtSeat(s, winner).name} wins the trick`;
  return s;
}

/** Clear the finished trick and start the next one (or end the hand). */
export function advanceAfterTrick(state: MendikotState): MendikotState {
  if (state.phase !== "trick_done") return state;
  const s = clone(state);
  const handsEmpty = s.players.every((p) => s.hands[p.id].length === 0);
  s.currentTrick = [];
  s.leadSuit = null;
  if (handsEmpty) {
    s.phase = "hand_over";
    s.result = computeResult(s);
    s.message = s.result.message;
    return s;
  }
  s.phase = "playing";
  s.leaderSeat = s.lastTrickWinnerSeat!;
  s.turnSeat = s.lastTrickWinnerSeat!;
  s.message = `${playerAtSeat(s, s.turnSeat).name} leads`;
  return s;
}

function computeResult(s: MendikotState): HandResult {
  const [tA, tB] = s.tensByTeam;
  const [trA, trB] = s.tricksByTeam;
  let winnerTeam: 0 | 1;
  let kind: HandResult["kind"];

  if (trA === 13 || trB === 13) {
    winnerTeam = trA === 13 ? 0 : 1;
    kind = "whitewash";
  } else if (tA === 4 || tB === 4) {
    winnerTeam = tA === 4 ? 0 : 1;
    kind = "mendikot";
  } else if (tA >= 3 || tB >= 3) {
    winnerTeam = tA >= 3 ? 0 : 1;
    kind = "tens";
  } else {
    // 2-2 split → more tricks wins
    winnerTeam = trA > trB ? 0 : 1;
    kind = "tricks";
  }

  const humanTeam = teamOf(HUMAN_SEAT);
  const humanWon = winnerTeam === humanTeam;
  const side = winnerTeam === humanTeam ? "Your team" : "Opponents";
  const messages: Record<HandResult["kind"], string> = {
    mendikot: `${side} swept all four 10s — Mendikot! 🎉`,
    whitewash: `${side} took all 13 tricks — Whitewash! 🧹`,
    tens: `${side} captured the majority of 10s.`,
    tricks: `10s split 2–2 — ${side} won on tricks.`,
  };
  return {
    winnerTeam,
    kind,
    tens: [tA, tB],
    tricks: [trA, trB],
    humanWon,
    message: messages[kind],
  };
}

/** Suit counts in a hand (for trump hints). */
export function suitCounts(hand: Card[]): Record<Suit, number> {
  const c: Record<Suit, number> = { S: 0, H: 0, D: 0, C: 0 };
  for (const card of hand) c[card.suit] += 1;
  return c;
}
